tmt/steps/prepare/__init__.py

```diff
@@ -22,40 +27,176 @@ def wake(self):
         self.super.wake()
 
         for i in range(len(self.data)):
-            self.set_default(i, 'how', 'shell')
-            self.set_default(i, 'playbooks', [])
-            self.set_default(i, 'path', self.data[i]['playbooks'])
-            self.set_default(i, 'script', self.data[i]['path'])
+            self.opts(i, 'how', 'script')
+
+            self.alias(i, 'script', 'playbooks')
+            self.alias(i, 'script', 'playbook')
+            self.alias(i, 'script', 'path')
+            self.alias(i, 'script', 'inline')
+
+            self.debug('how', self)
+
+            for key, val in self.data[i].items():
+                if not val is None:
+                    self.info(f'{key}', val)
 
     def show(self):
         """ Show discover details """
         self.super.show(keys = ['how', 'script'])
 
     def go(self):
         """ Prepare the test step """
+        # TODO: not sure why
         self.super.go()
 
+        packages = []
+        for step in self.plan.steps():
+            if getattr(step, 'requires', False):
+                packages += step.requires()
+
+        if packages:
+            self.install(packages)
+
         for data in self.data:
-            how = data['how']
-            script = data['script']
+            self.run_how(data['how'], data['script'])
```

Comment:
Code here traceback for me:

```
$ tmt run -da provision -h managed -r 10.29.162.240 -p fo0m4nchU
/var/tmp/tmt/run-050
Found 3 plans.

/plans/basic
summary: Essential command line features
        how: prepare
    how: ansible
    playbooks: plans/packages.yml
    name: one
    script: plans/packages.yml
        Read file '/var/tmp/tmt/run-050/plans/basic/discover/status.txt'.
        Write file '/var/tmp/tmt/run-050/plans/basic/discover/status.txt'.
    discover
        workdir: /var/tmp/tmt/run-050/plans/basic/discover
        how: fmf
        repository: https://github.com/psss/tmt
            Clone 'https://github.com/psss/tmt' to '/var/tmp/tmt/run-050/plans/basic/discover/one/tests'.
            Run command 'git clone https://github.com/psss/tmt /var/tmp/tmt/run-050/plans/basic/discover/one/tests'.
            err: Cloning into '/var/tmp/tmt/run-050/plans/basic/discover/one/tests'...
        revision: devel
            Checkout revision 'devel'.
            Run command 'git checkout -f devel'.
            err: Switched to a new branch 'devel'
            out: Branch 'devel' set up to track remote branch 'devel' from 'origin'.
        filter: tier: 0,1
            Check metadata tree in '/var/tmp/tmt/run-050/plans/basic/discover/one/tests'.
        tests: 2 tests selected
            /one/tests/docs
            /one/tests/ls
        Write file '/var/tmp/tmt/run-050/plans/basic/discover/steps.yaml'.
        Write file '/var/tmp/tmt/run-050/plans/basic/discover/tests.yaml'.
        Write file '/var/tmp/tmt/run-050/plans/basic/discover/status.txt'.
    provision
        workdir: /var/tmp/tmt/run-050/plans/basic/provision
        Provision dir: /var/tmp/tmt/run-050/plans/basic/provision/zQcKbnxYjvjyXsHx
            Run command 'vagrant version'.
            out: Installed Version: 2.2.5
            out
            out: Vagrant was unable to check for the latest version of Vagrant.
            out: Please check manually at https://www.vagrantup.com
            VagrantProvider: Checking initial status, setting defaults.
        how: managed
        name: one
        password: fo0m4nchU
        host: 10.29.162.240
        username: root
        box: fedora/31-cloud-base
        memory: 2048
            generating: managed
            Host: 10.29.162.240
            Run command 'bash -c "vagrant plugin list | grep '^vagrant-managed-servers '"'.
            out: vagrant-managed-servers (0.8.0, global)
            Run command '('vagrant', 'init', '-fm', 'tknerr/managed-server-dummy')'.
            out: A `Vagrantfile` has been placed in this directory. You are now
            out: ready to `vagrant up` your first virtual environment! Please read
            out: the comments in the Vagrantfile as well as documentation on
            out: `vagrantup.com` for more information on using Vagrant.
            Initialized new Vagrantfile: 
                                Vagrant.configure("2") do |config|
                                  config.vm.box = "tknerr/managed-server-dummy"
                                end
            Adding into Vagrantfile: vm.provider 'managed' do |managed| managed.server = "10.29.162.240"; end
            Run command 'vagrant validate'.
            out: Vagrantfile validated successfully.
            Adding into Vagrantfile: vm.synced_folder ".", "/vagrant", type: "rsync", disabled: true
            Run command 'vagrant validate'.
            out: Vagrantfile validated successfully.
            Adding into Vagrantfile: vm.synced_folder "/var/tmp/tmt/run-050/plans/basic", "/var/tmp/tmt/run-050/plans/basic", type: "rsync"
            Run command 'vagrant validate'.
            out: Vagrantfile validated successfully.
            Adding into Vagrantfile: ssh.username = "root"
            Run command 'vagrant validate'.
            out: Vagrantfile validated successfully.
            Adding into Vagrantfile: ssh.password = "fo0m4nchU"
            Run command 'vagrant validate'.
            out: Vagrantfile validated successfully.
            Adding into Vagrantfile: nfs.verify_installed = false
            Run command 'vagrant validate'.
            out: Vagrantfile validated successfully.
        Provisioning vagrant, Vagrantfile: 
                            Vagrant.configure("2") do |config|
                              config.vm.box = "tknerr/managed-server-dummy"
                              config.vm.provider 'managed' do |managed| managed.server = "10.29.162.240"; end
                              config.vm.synced_folder ".", "/vagrant", type: "rsync", disabled: true
                              config.vm.synced_folder "/var/tmp/tmt/run-050/plans/basic", "/var/tmp/tmt/run-050/plans/basic", type: "rsync"
                              config.ssh.username = "root"
                              config.ssh.password = "fo0m4nchU"
                              config.nfs.verify_installed = false
                            end
            Run command 'vagrant up'.
            out: Bringing machine 'default' up with 'managed' provider...
            out: ==> default: Linking with managed server 10.29.162.240
            out: ==> default:  -- Server: 10.29.162.240
    prepare
        workdir: /var/tmp/tmt/run-050/plans/basic/prepare
            Adding into Vagrantfile: vm.provision "prepare", type: "shell", privileged: "true", run: "never", inline: "set -o pipefail; dnf install -y beakerlib 2>&1 | tee -a '/var/tmp/tmt/run-050/plans/basic/prepare/prepare.log'"
            Run command 'vagrant validate'.



            out:     default: 
            out:     default: Inserting generated public key within guest...
            out:     default: Removing insecure key from the guest if it's present...
            out:     default: Key inserted! Disconnecting and reconnecting using new SSH key...
            out: Vagrantfile validated successfully.
            Run command '('vagrant', 'provision', '--provision-with', 'prepare')'.





            out: ==> default: Rsyncing folder: /var/tmp/tmt/run-050/plans/basic/ => /var/tmp/tmt/run-050/plans/basic



            out: ==> default: Running provisioner: prepare (shell)...
            out:     default: Running: inline script
            out:     default: Updating Subscription Management repositories.






            out:     default: Unable to read consumer identity
            out:     default: This system is not registered to Red Hat Subscription Management. You can use subscription-manager to register.
            out:     default: Copr repo for beakerlib-libraries owned by bgon 0.0  B/s |   0  B     02:00    
            out:     default: Errors during downloading metadata for repository 'beakerlib-libraries':
            out:     default:   - Curl error (28): Timeout was reached for https://coprbe.devel.redhat.com/results/bgoncalv/beakerlib-libraries/rhel-8.dev-x86_64/repodata/repomd.xml [Connection timed out after 30000 milliseconds]
            out:     default:   - Curl error (28): Timeout was reached for https://coprbe.devel.redhat.com/results/bgoncalv/beakerlib-libraries/rhel-8.dev-x86_64/repodata/repomd.xml [Connection timed out after 30001 milliseconds]
            out:     default: Error: Failed to download metadata for repo 'beakerlib-libraries': Cannot download repomd.xml: Cannot download repodata/repomd.xml: All mirrors were tried
            out:     default: Ignoring repositories: beakerlib-libraries
            out:     default: Last metadata expiration check: 0:15:47 ago on Mon 09 Dec 2019 04:43:59 PM EST.
            out:     default: Package beakerlib-1.18-6.el8bkr.noarch is already installed.
            out:     default: Dependencies resolved.
            out:     default: Nothing to do.
            out:     default: Complete!
            Run command 'bash -c "vagrant plugin list | grep '^vagrant-rsync-back '"'.
            out: vagrant-rsync-back (0.0.1, global)
            Run command 'vagrant rsync-back'.
            out: ==> default: Rsyncing folder: /var/tmp/tmt/run-050/plans/basic/ => /var/tmp/tmt/run-050/plans/basic
        /var/tmp/tmt/run-050/plans/basic/prepare/prepare.log: 
            Updating Subscription Management repositories.
            Unable to read consumer identity
            This system is not registered to Red Hat Subscription Management. You can use subscription-manager to register.
            Copr repo for beakerlib-libraries owned by bgon 0.0  B/s |   0  B     02:00    
            Errors during downloading metadata for repository 'beakerlib-libraries':
              - Curl error (28): Timeout was reached for https://coprbe.devel.redhat.com/results/bgoncalv/beakerlib-libraries/rhel-8.dev-x86_64/repodata/repomd.xml [Connection timed out after 30000 milliseconds]
              - Curl error (28): Timeout was reached for https://coprbe.devel.redhat.com/results/bgoncalv/beakerlib-libraries/rhel-8.dev-x86_64/repodata/repomd.xml [Connection timed out after 30001 milliseconds]
            Error: Failed to download metadata for repo 'beakerlib-libraries': Cannot download repomd.xml: Cannot download repodata/repomd.xml: All mirrors were tried
            Ignoring repositories: beakerlib-libraries
            Last metadata expiration check: 0:15:47 ago on Mon 09 Dec 2019 04:43:59 PM EST.
            Package beakerlib-1.18-6.el8bkr.noarch is already installed.
            Dependencies resolved.
            Nothing to do.
            Complete!
    Prepare: ansible = 'plans/packages.yml
Traceback (most recent call last):
  File "/home/mvadkert/venvs/tmt/bin/tmt", line 7, in <module>
    exec(compile(f.read(), __file__, 'exec'))
  File "/home/mvadkert/git/github.com/tmt/bin/tmt", line 11, in <module>
    tmt.cli.main()
  File "/home/mvadkert/venvs/tmt/lib/python3.7/site-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/home/mvadkert/venvs/tmt/lib/python3.7/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/home/mvadkert/venvs/tmt/lib/python3.7/site-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/mvadkert/venvs/tmt/lib/python3.7/site-packages/click/core.py", line 1164, in invoke
    return _process_result(rv)
  File "/home/mvadkert/venvs/tmt/lib/python3.7/site-packages/click/core.py", line 1102, in _process_result
    **ctx.params)
  File "/home/mvadkert/venvs/tmt/lib/python3.7/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/home/mvadkert/venvs/tmt/lib/python3.7/site-packages/click/decorators.py", line 17, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/home/mvadkert/git/github.com/tmt/tmt/cli.py", line 330, in finito
    context.obj.run.go()
  File "/home/mvadkert/git/github.com/tmt/tmt/base.py", line 553, in go
    plan.go()
  File "/home/mvadkert/git/github.com/tmt/tmt/base.py", line 310, in go
    step.go()
  File "/home/mvadkert/git/github.com/tmt/tmt/steps/prepare/__init__.py", line 61, in go
    self.run_how(data['how'], data['script'])
  File "/home/mvadkert/git/github.com/tmt/tmt/steps/prepare/__init__.py", line 71, in run_how
    )(how, what)
TypeError: <lambda>() takes 0 positional arguments but 2 were given
```

Reply:
Well that's strange:
```
            Unable to read consumer identity
            This system is not registered to Red Hat Subscription Management. You can use subscription-manager to register.
            Copr repo for beakerlib-libraries owned by bgon 0.0  B/s |   0  B     02:00    
            Errors during downloading metadata for repository 'beakerlib-libraries':
              - Curl error (28): Timeout was reached for https://coprbe.devel.redhat.com/results/bgoncalv/beakerlib-libraries/rhel-8.dev-x86_64/repodata/repomd.xml [Connection timed out after 30000 milliseconds]
              - Curl error (28): Timeout was reached for https://coprbe.devel.redhat.com/results/bgoncalv/beakerlib-libraries/rhel-8.dev-x86_64/repodata/repomd.xml [Connection timed out after 30001 milliseconds]
            Error: Failed to download metadata for repo 'beakerlib-libraries': Cannot download repomd.xml: Cannot download repodata/repomd.xml: All mirrors were tried
            Ignoring repositories: beakerlib-libraries
            Last metadata expiration check: 0:15:47 ago on Mon 09 Dec 2019 04:43:59 PM EST.
            Package beakerlib-1.18-6.el8bkr.noarch is already installed.
```
Anyway, the error is mine, literally `def how_generic()`.  I'll account for those.

Reply:
This error is expected, I run in on an AWS machine and we do not have yet access to coprbe.devel.redhat.com

Reply:
@pvalena this fixes the issue:

```
diff --git a/tmt/steps/prepare/__init__.py b/tmt/steps/prepare/__init__.py
index 1925be1..214aeb4 100644
--- a/tmt/steps/prepare/__init__.py
+++ b/tmt/steps/prepare/__init__.py
@@ -65,10 +65,7 @@ class Prepare(tmt.steps.Step):
     def run_how(self, how, what):
         """ Run specific HOW """
         self.info('Prepare', f"{how} = '{what}", 'yellow')
-        getattr(self,
-            f"how_{how}",
-            lambda: 'generic',
-            )(how, what)
+        getattr(self, f"how_{how}", self.how_generic)(how, what)
 
     def how_generic(self, how, what):
         if not what:
```

Reply:
submitted as PR to your fork